File: internal/store/store_test.go
package store

import (
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
	"time"
)

// Unit tests for the learning store: canonicalisation, recency-weighted
// ranking, deploy presets, and a round-trip to a TEMP dir.
//
// Every persistence test injects an explicit temp BaseDir — real ~/.kc is never
// written.
// Ported from tools/kc-bun/test/store.test.ts.

var scope = Scope{Cluster: "thinkpilot-k3s", App: "mailon"}

// ── Pure helpers ────────────────────────────────────────────────────────────

func TestCanonicalKey(t *testing.T) {
	if CanonicalKey(map[string]any{"a": 1, "b": 2}) != CanonicalKey(map[string]any{"b": 2, "a": 1}) {
		t.Error("key order should not matter")
	}
	if CanonicalKey(map[string]any{"x": map[string]any{"p": 1, "q": 2}, "list": []any{1, 2}}) !=
		CanonicalKey(map[string]any{"list": []any{1, 2}, "x": map[string]any{"q": 2, "p": 1}}) {
		t.Error("nested objects/arrays should normalise")
	}
	if CanonicalKey(map[string]any{"deployments": []any{"web"}}) ==
		CanonicalKey(map[string]any{"deployments": []any{"web", "sender"}}) {
		t.Error("distinct values should differ")
	}
}

func TestNormalizeSet(t *testing.T) {
	got := NormalizeSet([]string{" web ", "sender", "web", "", "sender"})
	if !reflect.DeepEqual(got, []string{"sender", "web"}) {
		t.Errorf("NormalizeSet = %v, want [sender web]", got)
	}
}

func TestRankParams_FrequencyAndRecency(t *testing.T) {
	const now int64 = 1_000_000_000_000
	const day int64 = 24 * 60 * 60 * 1000

	t.Run("more frequent ranks higher", func(t *testing.T) {
		records := []ActionRecord{
			{Action: "deploy", Params: Params{"d": []any{"web"}}, TS: now - day},
			{Action: "deploy", Params: Params{"d": []any{"web"}}, TS: now - 2*day},
			{Action: "deploy", Params: Params{"d": []any{"sender"}}, TS: now - day},
		}
		ranked := RankParams(records, RankOptions{Now: now})
		if len(ranked) != 2 {
			t.Fatalf("got %d distinct, want 2", len(ranked))
		}
		if !reflect.DeepEqual(ranked[0]["d"], []any{"web"}) {
			t.Errorf("top = %v, want web", ranked[0]["d"])
		}
	})

	t.Run("recency outweighs old single with short half-life", func(t *testing.T) {
		records := []ActionRecord{
			{Action: "deploy", Params: Params{"d": []any{"old"}}, TS: now - 60*day},
			{Action: "deploy", Params: Params{"d": []any{"fresh"}}, TS: now},
		}
		ranked := RankParams(records, RankOptions{Now: now, HalfLife: day})
		if !reflect.DeepEqual(ranked[0]["d"], []any{"fresh"}) {
			t.Errorf("top = %v, want fresh", ranked[0]["d"])
		}
	})

	t.Run("ties broken by most-recent occurrence", func(t *testing.T) {
		records := []ActionRecord{
			{Action: "deploy", Params: Params{"d": []any{"a"}}, TS: now - 10*day},
			{Action: "deploy", Params: Params{"d": []any{"b"}}, TS: now - 1*day},
		}
		ranked := RankParams(records, RankOptions{Now: now, HalfLife: 10_000 * day})
		if !reflect.DeepEqual(ranked[0]["d"], []any{"b"}) {
			t.Errorf("top = %v, want b (more recent)", ranked[0]["d"])
		}
	})
}

// ── Persistence round-trip (TEMP dir) ───────────────────────────────────────

func TestActionHistory_NeverResolvesRealHomeWhenBaseDirInjected(t *testing.T) {
	dir := t.TempDir()
	h := New(Options{BaseDir: dir})
	if !strings.HasPrefix(h.Path(), dir) {
		t.Errorf("path %q does not start with temp dir %q", h.Path(), dir)
	}
	home, _ := os.UserHomeDir()
	if strings.Contains(h.Path(), filepath.Join(home, ".kc")) {
		t.Errorf("path %q must not touch real ~/.kc", h.Path())
	}
}

func TestActionHistory_RoundTrip(t *testing.T) {
	dir := t.TempDir()
	const t0 int64 = 1_700_000_000_000
	write := New(Options{BaseDir: dir, Now: func() int64 { return t0 }})
	if err := write.Record("deploy", scope, Params{"deployments": []any{"web"}}); err != nil {
		t.Fatalf("record: %v", err)
	}
	if _, err := os.Stat(write.Path()); err != nil {
		t.Fatalf("state file not written: %v", err)
	}
	// A brand-new instance reads the same persisted state.
	read := New(Options{BaseDir: dir})
	ranked := read.Ranked("deploy", scope)
	if len(ranked) != 1 || !reflect.DeepEqual(ranked[0]["deployments"], []any{"web"}) {
		t.Errorf("reloaded ranked = %v, want [{deployments:[web]}]", ranked)
	}
}

func TestActionHistory_ScopeIsolation(t *testing.T) {
	dir := t.TempDir()
	h := New(Options{BaseDir: dir})
	_ = h.RecordDeploy(Scope{Cluster: "c1", App: "mailon"}, []string{"web"})
	_ = h.RecordDeploy(Scope{Cluster: "c1", App: "other"}, []string{"api"})
	_ = h.RecordDeploy(Scope{Cluster: "c2", App: "mailon"}, []string{"sender"})

	check := func(s Scope, want [][]string) {
		t.Helper()
		if got := h.DeployPresets(s); !reflect.DeepEqual(got, want) {
			t.Errorf("DeployPresets(%+v) = %v, want %v", s, got, want)
		}
	}
	check(Scope{Cluster: "c1", App: "mailon"}, [][]string{{"web"}})
	check(Scope{Cluster: "c1", App: "other"}, [][]string{{"api"}})
	check(Scope{Cluster: "c2", App: "mailon"}, [][]string{{"sender"}})
}

func TestActionHistory_MissingStateFile(t *testing.T) {
	h := New(Options{BaseDir: filepath.Join(t.TempDir(), "does-not-exist-yet")})
	if got := h.Ranked("deploy", scope); len(got) != 0 {
		t.Errorf("ranked = %v, want empty", got)
	}
	if got := h.DeployPresets(scope); len(got) != 0 {
		t.Errorf("presets = %v, want empty", got)
	}
}

func TestActionHistory_CorruptStateFile(t *testing.T) {
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "state.json"), []byte("{ not valid json "), 0o644); err != nil {
		t.Fatal(err)
	}
	h := New(Options{BaseDir: dir})
	if got := h.Ranked("deploy", scope); len(got) != 0 {
		t.Errorf("ranked = %v, want empty (corrupt → clean start)", got)
	}
}

// ── Deploy presets — the spec's headline scenario ───────────────────────────

func TestDeployPresets_TwoPresetsMostRecentFirst(t *testing.T) {
	dir := t.TempDir()
	clock := int64(1_700_000_000_000)
	h := New(Options{BaseDir: dir, Now: func() int64 { return clock }})

	clock += 1000
	_ = h.RecordDeploy(scope, []string{"web"})
	clock += 1000
	_ = h.RecordDeploy(scope, []string{"responder", "sender"})

	presets := h.DeployPresets(scope)
	if len(presets) != 2 {
		t.Fatalf("got %d presets, want 2", len(presets))
	}
	if !reflect.DeepEqual(presets[0], []string{"responder", "sender"}) {
		t.Errorf("preset[0] = %v, want [responder sender] (most recent)", presets[0])
	}
	if !reflect.DeepEqual(presets[1], []string{"web"}) {
		t.Errorf("preset[1] = %v, want [web]", presets[1])
	}
}

func TestDeployPresets_MostRecentFirstWhenTimestampsTie(t *testing.T) {
	// Two deploys within the same millisecond → identical ts. Insertion order
	// (the lastIndex tie-break) must still decide "most recent".
	dir := t.TempDir()
	const frozen int64 = 1_700_000_000_000
	h := New(Options{BaseDir: dir, Now: func() int64 { return frozen }})
	_ = h.RecordDeploy(scope, []string{"web"})
	_ = h.RecordDeploy(scope, []string{"responder", "sender"})

	presets := h.DeployPresets(scope)
	if !reflect.DeepEqual(presets[0], []string{"responder", "sender"}) {
		t.Errorf("preset[0] = %v, want [responder sender]", presets[0])
	}
	if !reflect.DeepEqual(presets[1], []string{"web"}) {
		t.Errorf("preset[1] = %v, want [web]", presets[1])
	}
}

func TestDeployPresets_RedeploySameSetRanksNotCounts(t *testing.T) {
	dir := t.TempDir()
	clock := int64(1_700_000_000_000)
	h := New(Options{BaseDir: dir, Now: func() int64 { return clock }})

	// [web] deployed twice, [sender] once but more recently.
	clock += 1000
	_ = h.RecordDeploy(scope, []string{"web"})
	clock += 1000
	_ = h.RecordDeploy(scope, []string{"web"})
	clock += 1000
	_ = h.RecordDeploy(scope, []string{"sender"})

	presets := h.DeployPresets(scope)
	if len(presets) != 2 {
		t.Fatalf("got %d presets, want 2", len(presets))
	}
	// [web] has higher frequency → ranks first under the default half-life.
	if !reflect.DeepEqual(presets[0], []string{"web"}) {
		t.Errorf("preset[0] = %v, want [web] (higher frequency)", presets[0])
	}

	// Order-insensitive: [sender,responder] then [responder,sender] is the same
	// permutation.
	clock += 1000
	_ = h.RecordDeploy(scope, []string{"sender", "responder"})
	clock += 1000
	_ = h.RecordDeploy(scope, []string{"responder", "sender"})
	presets2 := h.DeployPresets(scope)
	count := 0
	for _, p := range presets2 {
		if len(p) == 2 && p[0] == "responder" && p[1] == "sender" {
			count++
		}
	}
	if count != 1 {
		t.Errorf("[responder sender] permutation appears %d times, want 1", count)
	}
}

// ── Default half-life sanity (uses real DefaultHalfLife) ─────────────────────

func TestRankParams_DefaultHalfLife(t *testing.T) {
	now := time.Now().UnixMilli()
	day := int64(24 * time.Hour / time.Millisecond)
	records := []ActionRecord{
		{Action: "deploy", Params: Params{"d": []any{"recent"}}, TS: now - day},
		{Action: "deploy", Params: Params{"d": []any{"ancient"}}, TS: now - 365*day},
	}
	ranked := RankParams(records, RankOptions{Now: now}) // default half-life
	if !reflect.DeepEqual(ranked[0]["d"], []any{"recent"}) {
		t.Errorf("top = %v, want recent under the default 2-week half-life", ranked[0]["d"])
	}
}

File: test/store.test.ts
/**
 * Unit tests for the learning store: canonicalisation, recency-weighted ranking,
 * deploy presets, and a round-trip to a TEMP dir.
 *
 * Every test injects an explicit temp baseDir — real `~/.kc` is never written.
 */

import { test, expect, describe, beforeEach, afterEach } from "bun:test"
import { mkdtemp, rm } from "node:fs/promises"
import { tmpdir, homedir } from "node:os"
import { join } from "node:path"
import {
  ActionHistory,
  rankParams,
  canonicalKey,
  normalizeSet,
} from "../src/store/index.ts"
import type { ActionRecord, Scope } from "../src/types.ts"

const SCOPE: Scope = { cluster: "thinkpilot-k3s", app: "mailon" }

// ── Pure helpers ────────────────────────────────────────────────────────────

describe("canonicalKey", () => {
  test("key order independent", () => {
    expect(canonicalKey({ a: 1, b: 2 })).toBe(canonicalKey({ b: 2, a: 1 }))
  })
  test("nested objects/arrays normalised", () => {
    expect(canonicalKey({ x: { p: 1, q: 2 }, list: [1, 2] })).toBe(
      canonicalKey({ list: [1, 2], x: { q: 2, p: 1 } }),
    )
  })
  test("distinct values differ", () => {
    expect(canonicalKey({ deployments: ["web"] })).not.toBe(
      canonicalKey({ deployments: ["web", "sender"] }),
    )
  })
})

describe("normalizeSet", () => {
  test("dedupes, trims, sorts", () => {
    expect(normalizeSet([" web ", "sender", "web", "", "sender"])).toEqual(["sender", "web"])
  })
})

describe("rankParams (recency-weighted frequency)", () => {
  const now = 1_000_000_000_000
  const day = 24 * 60 * 60 * 1000

  test("more frequent params rank higher", () => {
    const records: ActionRecord[] = [
      { action: "deploy", params: { d: ["web"] }, ts: now - day },
      { action: "deploy", params: { d: ["web"] }, ts: now - 2 * day },
      { action: "deploy", params: { d: ["sender"] }, ts: now - day },
    ]
    const ranked = rankParams<{ d: string[] }>(records, { now })
    expect(ranked[0]!.d).toEqual(["web"])
    expect(ranked.length).toBe(2)
  })

  test("recency outweighs a single old occurrence with a short half-life", () => {
    const records: ActionRecord[] = [
      // 'old' deployed once long ago.
      { action: "deploy", params: { d: ["old"] }, ts: now - 60 * day },
      // 'fresh' deployed once just now.
      { action: "deploy", params: { d: ["fresh"] }, ts: now },
    ]
    const ranked = rankParams<{ d: string[] }>(records, { now, halfLifeMs: day })
    expect(ranked[0]!.d).toEqual(["fresh"])
  })

  test("ties broken by most-recent occurrence", () => {
    const records: ActionRecord[] = [
      { action: "deploy", params: { d: ["a"] }, ts: now - 10 * day },
      { action: "deploy", params: { d: ["b"] }, ts: now - 1 * day },
    ]
    // With a very long half-life both score ~1; the more recent ('b') wins.
    const ranked = rankParams<{ d: string[] }>(records, {
      now,
      halfLifeMs: 10_000 * day,
    })
    expect(ranked[0]!.d).toEqual(["b"])
  })
})

// ── Persistence round-trip (TEMP dir) ───────────────────────────────────────

describe("ActionHistory persistence", () => {
  let dir: string

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "kc-store-test-"))
  })
  afterEach(async () => {
    await rm(dir, { recursive: true, force: true })
  })

  test("never resolves to the real ~/.kc when baseDir is injected", () => {
    const store = new ActionHistory({ baseDir: dir })
    expect(store.path.startsWith(dir)).toBe(true)
    expect(store.path).not.toContain(join(homedir(), ".kc"))
  })

  test("round-trips records to disk and reloads them in a fresh instance", async () => {
    const t0 = 1_700_000_000_000
    const write = new ActionHistory({ baseDir: dir, now: () => t0 })
    await write.record("deploy", SCOPE, { deployments: ["web"] })

    // The state file exists under the temp dir.
    expect(await Bun.file(write.path).exists()).toBe(true)

    // A brand-new instance reads the same persisted state.
    const read = new ActionHistory({ baseDir: dir })
    const ranked = await read.ranked("deploy", SCOPE)
    expect(ranked).toEqual([{ deployments: ["web"] }])
  })

  test("scopes are isolated by cluster × app", async () => {
    const store = new ActionHistory({ baseDir: dir })
    await store.recordDeploy({ cluster: "c1", app: "mailon" }, ["web"])
    await store.recordDeploy({ cluster: "c1", app: "other" }, ["api"])
    await store.recordDeploy({ cluster: "c2", app: "mailon" }, ["sender"])

    expect(await store.deployPresets({ cluster: "c1", app: "mailon" })).toEqual([["web"]])
    expect(await store.deployPresets({ cluster: "c1", app: "other" })).toEqual([["api"]])
    expect(await store.deployPresets({ cluster: "c2", app: "mailon" })).toEqual([["sender"]])
  })

  test("missing state file → empty rankings (no crash)", async () => {
    const store = new ActionHistory({ baseDir: join(dir, "does-not-exist-yet") })
    expect(await store.ranked("deploy", SCOPE)).toEqual([])
    expect(await store.deployPresets(SCOPE)).toEqual([])
  })

  test("corrupt state file → starts clean", async () => {
    await Bun.write(join(dir, "state.json"), "{ not valid json ")
    const store = new ActionHistory({ baseDir: dir })
    expect(await store.ranked("deploy", SCOPE)).toEqual([])
  })
})

// ── Deploy presets — the spec's headline scenario ───────────────────────────

describe("deploy presets (SPEC scenario)", () => {
  let dir: string
  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "kc-presets-"))
  })
  afterEach(async () => {
    await rm(dir, { recursive: true, force: true })
  })

  test("record [web] then [responder,sender] → two presets, most-recent first", async () => {
    let clock = 1_700_000_000_000
    const store = new ActionHistory({ baseDir: dir, now: () => clock })

    clock += 1000
    await store.recordDeploy(SCOPE, ["web"])
    clock += 1000
    await store.recordDeploy(SCOPE, ["responder", "sender"])

    const presets = await store.deployPresets(SCOPE)
    expect(presets.length).toBe(2)
    // Most-recent first (recency-weighted; equal frequency → newer wins).
    expect(presets[0]).toEqual(["responder", "sender"]) // normalized (sorted) set
    expect(presets[1]).toEqual(["web"])
  })

  test("most-recent-first holds even when both records share a timestamp", async () => {
    // Real-world case: two deploys within the same millisecond → identical ts.
    // Insertion order must still decide "most recent" (the lastIndex tie-break).
    const frozen = 1_700_000_000_000
    const store = new ActionHistory({ baseDir: dir, now: () => frozen })
    await store.recordDeploy(SCOPE, ["web"])
    await store.recordDeploy(SCOPE, ["responder", "sender"])

    const presets = await store.deployPresets(SCOPE)
    expect(presets[0]).toEqual(["responder", "sender"])
    expect(presets[1]).toEqual(["web"])
  })

  test("re-deploying the same set increases its rank, not its count of presets", async () => {
    let clock = 1_700_000_000_000
    const store = new ActionHistory({ baseDir: dir, now: () => clock })

    // [web] deployed twice, [sender] once — but [sender] more recently.
    clock += 1000
    await store.recordDeploy(SCOPE, ["web"])
    clock += 1000
    await store.recordDeploy(SCOPE, ["web"])
    clock += 1000
    await store.recordDeploy(SCOPE, ["sender"])

    const presets = await store.deployPresets(SCOPE)
    // Still two distinct permutations.
    expect(presets.length).toBe(2)
    // [web] has higher frequency → ranks first under the default half-life.
    expect(presets[0]).toEqual(["web"])

    // Order-insensitive: deploying [sender,responder] then [responder,sender]
    // is the same permutation.
    clock += 1000
    await store.recordDeploy(SCOPE, ["sender", "responder"])
    clock += 1000
    await store.recordDeploy(SCOPE, ["responder", "sender"])
    const presets2 = await store.deployPresets(SCOPE)
    const permutation = presets2.filter(
      (p) => p.length === 2 && p[0] === "responder" && p[1] === "sender",
    )
    expect(permutation.length).toBe(1)
  })
})
